fix: Don't add indices on columns with UNIQUE constraint

hidden-primary-net opened the pull request:
SQL::Translator::Parser::DBIx::Class adds an index to foreign key
columns as this "is normally the sensible thing to do". Agreed.

Besides SQL::Translator::Parser::DBIx::Class takes care to not add an
additional index to primary key columns ("some RDBMS croak on this, and
it generally doesn't make much sense") but doesn't consider columns with
a UNIQUE constraint where it doesn't make any sense either.

Generally this should not be dangerous but it doesn't help at all.
From my understanding the SQL optimizer will use the better index thus
the UNIQUE constraint would win in any case.

Even worse any INSERT, UPDATE or DELETE operation would need to maintain
the index which is never used. This does not sound sensible to me.

The supposed patch avoids the adding of those indices if the column is
checked if it is *the first column* of a UNIQUE constraint. If it's not
the index is added, if it is the first column it is handled like a
primary key.

hidden-primary-net committed:
fix: Don't add indices on columns with UNIQUE constraint

SQL::Translator::Parser::DBIx::Class adds an index to foreign key
columns as this "is normally the sensible thing to do". Agreed.

Besides SQL::Translator::Parser::DBIx::Class takes care to not add an
additional index to primary key columns ("some RDBMS croak on this, and
it generally doesn't make much sense") but doesn't consider columns with
a UNIQUE constraint where it doesn't make any sense either.

Generally this should not be dangerous but it doesn't help at all.
From my understanding the SQL optimizer will use the better index thus
the UNIQUE constraint would win in any case.

Even worse any INSERT, UPDATE or DELETE operation would need to maintain
the index which is never used. This does not sound sensible to me.

The supposed patch avoids the adding of those indices if the column is
checked if it is *the first column* of a UNIQUE constraint. If it's not
the index is added, if it is the first column it is handled like a
primary key.

msouth commented:
Another consideration in favor of this is the disk space that is taken up by an unnecessary index.

BTW, is something wrong with the Travis setup?  I tried to look into the supposed failure and the one I picked that said it failed looked like it was displaying the output of a successful build ("all tests passed", and "job exited with 0", etc in the output).  The one I looked at was https://travis-ci.org/dbsrgits/dbix-class/jobs/198374714

hidden-primary-net committed:
test: Add test for redundant indices

I want to achieve a test checking the given schema for automatic indices
on foreign key columns actually being covered by a UNIQUE constraint.

Unfortunately I don't understand the difference between this test and
t/86sqlt.

* I have to "use SQL::Translator"
* tables are not populated in SQL::Translator->schema
* SQL::Translator->schema->get_table returns undef

hidden-primary-net committed:
test: Add build dependency

Test were failing earlier, thus adding SQL::Translator build dep.
